File: etc/fibonacci.js
//// brute force
const fib = n => {
    let arr = [0, 1]
    for(let i = 2; i <= n; i++) {
        arr.push(arr[i - 1] + arr[i - 2])
    }
    return arr[n]
}

//// recursion
// const fib = n => {
//     if(n < 2) return n
//     return fib(n - 1) + fib(n - 2)
// }

//// recursion + memoization 
// const cache = {}
// const fib = n => {
//     if(n < 2) return n
//     if(cache[n]) return cache[n]
//     let result = fib(n - 1) + fib(n - 2)
//     cache[n] = result
//     return result
// }

console.log(fib(10))
console.log(fib(20))
console.log(fib(30))
console.log(fib(40))
console.log(fib(50))
// console.log(fib(75))
// console.log(fib(78))

File: etc/fibonacci.py
cache = {}

def fib(n):
  if n < 2:
    return n
  if n in cache.keys():
    return cache[n]
  result = fib(n - 1) + fib(n - 2)
  cache[n] = result
  return result

print(fib(0)) # 0
print(fib(1)) # 1
print(fib(2)) # 1
print(fib(3)) # 2
print(fib(4)) # 3
print(fib(5)) # 5
print(fib(6)) # 8
print(fib(7)) # 13
print(fib(8)) # 21
print(fib(9)) # 34
print(fib(10)) # 55
print(fib(100)) # ?
print(fib(1000)) # ?
print(fib(1997)) # ?